gr-fec/lib/fec_mtrx_impl.cc

```diff
@@ -277,100 +273,111 @@ fec_mtrx_impl::fec_mtrx_impl()
+            gsl_matrix_uchar_set(result, row_index, col_index, (value1 + value2) & 1);
         }
     }
 }
 
-void fec_mtrx_impl::mult_matrices_mod2(gsl_matrix* result,
-                                       const gsl_matrix* matrix1,
-                                       const gsl_matrix* matrix2) const
+void fec_mtrx_impl::mult_matrices_mod2(gsl_matrix_uchar* result,
+                                       const gsl_matrix_uchar* matrix1,
+                                       const gsl_matrix_uchar* matrix2) const
 {
     // Verify that matrix sizes are appropriate
-    unsigned int a = (*matrix1).size1; // # of rows
-    unsigned int b = (*matrix1).size2; // # of columns
-    unsigned int c = (*matrix2).size1; // # of rows
-    unsigned int d = (*matrix2).size2; // # of columns
-    if (b != c) {
-        std::cout << "Error in "
-                  << "fec_mtrx_impl::mult_matrices_mod2."
-                  << " Matrix dimensions do not allow for matrix "
-                  << "multiplication operation:\nmatrix1 is " << a << " x " << b
-                  << ", and matrix2 is " << c << " x " << d << ".\n";
-        throw std::runtime_error("fec_mtrx::mult_matrices_mod2");
+    const unsigned int n_rows1 = matrix1->size1; // # of rows
+    const unsigned int n_cols1 = matrix1->size2; // # of columns
+    const unsigned int n_rows2 = matrix2->size1; // # of rows
+    const unsigned int n_cols2 = matrix2->size2; // # of columnsa
+
+    const unsigned int n_rows_res = n_rows1;
+    const unsigned int n_cols_res = n_cols2;
+
+    if (n_cols1 != n_rows2) {
+        d_logger.error("mult_matrices_mod2: Matrix dimensions do not allow for matrix "
+                       "multiplication operation:\n"
+                       "matrix1 is {} x {}, and matrix2 is {} x {}",
+                       n_rows1,
+                       n_cols1,
+                       n_rows2,
+                       n_cols2);
+        throw std::runtime_error("fec_mtrx::mult_matrices_mod2: size mismatch");
     }
 
-    // Perform matrix multiplication. This is not mod 2.
-    gsl_blas_dgemm(CblasNoTrans, CblasNoTrans, 1.0, matrix1, matrix2, 0.0, result);
-
-    // Take care of mod 2 manually.
-    unsigned int row_index, col_index;
-    unsigned int rows = (*result).size1, cols = (*result).size2;
-    for (row_index = 0; row_index < rows; row_index++) {
-        for (col_index = 0; col_index < cols; col_index++) {
-            int value = gsl_matrix_get(result, row_index, col_index);
-            int new_value = value % 2;
-            gsl_matrix_set(result, row_index, col_index, new_value);
+    for (unsigned int row = 0; row < n_rows_res; row++) {
+        for (unsigned int col = 0; col < n_cols_res; col++) {
+            unsigned char sum = 0;
+            for (unsigned int idx = 0; idx < n_cols1; idx++) {
+                sum += gsl_matrix_uchar_get(matrix1, row, idx) *
+                       gsl_matrix_uchar_get(matrix2, idx, col);
```

Comment:
I would suggest some care here, since `sum` can overflow. Probably this is not going to happen for LDPC matrices, because they have few 1's (that what "LD" means), but it can happen for dense generator matrices (I'm not really paying attention to which kinds of matrices actually need to be passed to this function; I'd rather have a multiply function that works for all GF(2) matrices unless they are very large). So I would use a `uint32_t` at least for `sum`.

Reply:
I think since all that matters in the end is `sum mod 2`, that's not a problem?

Reply:
You are right. But then, why don't we write this as the correct operations in GF(2)? Like so:
```c++
                sum ^= gsl_matrix_uchar_get(matrix1, row, idx) &
                       gsl_matrix_uchar_get(matrix2, idx, col);
```
I don't know if this allows the compiler to perform additional optimizations, but to me just the fact that this makes the intent of operating in GF(2) clear makes this approach preferable.